essential-scala/structural-recursion.md

```diff
@@ -1,6 +1,198 @@
 # Structural Recursion
 
-Core concept: transformation of an algebraic data type
+## Overview {:teacher:}
+
+Core concept: transformation of an algebraic data type. The structure of the transform follows the structure of the data.
+
 Scala features:
 - pattern matching
 - polymorphism
+
+Rule: any time you need to transform an algebraic data type you can implement this transform using structural recursion.
+
+
+## Introduction {:present:}
+
+After the previous lesson you now understand ADT and how to model your domain using them. But, of course, we don't want just to model a domain. We want to do things with it, and get results operating with these ADT.
+
+In today's lesson we will start looking at *polymorphism*, which helps us work with our model while we avoid duplication in our code. Later we will look at *pattern matching*, a way to interact with an ADT based on the shape of the data.
+
+{:teacher:} beware parametric polymorphism, it can be harded to understand than it seems, depending on previous programming experience. Don't introduce implicits here, it may confuse the issue more.
+
+## Polymorphism
```

Comment:
I don't think a lot of the material on polymorphism is necessary. When we teach structural recursion the points we make are:

1. Any transformation on an algebraic data type can be implemented using structural recursion.
2. There are two ways to implement structural recursion:

* polymorphism; or
* pattern matching

3. Given an algebraic data type, implementing the structural recursion skeleton is formulaic for either implementation method.

If `A` is a `B` or `C` (sum type) for pattern matching the skeleton for pattern matching is

```scala
sealed trait A {
  def someMethod: SomeType =
    this match {
      case B => ???
      case C => ???
    }
}
final case class B() extends A
final case class C() extends A
```

and for polymorphism is

```scala
sealed trait A {
  def someMethod: SomeType
}
final case class B() extends A {
  def someMethod: SomeType = ???
}
final case class C() extends A {
  def someMethod: SomeType = ???
}
```

(and so on for product types, which should be straightforward).


4. There is one final rule: if the data is recursive so is the method

The point of all this is that you can write this code in a completely systematic and repeatable way, just like you can implement algebraic data types in a completely systematic and repeatable way---the code follows directly from the structure of the data. Fix the structure of the data and both the code defining the data and the skeleton of any code to transform that data follows *immediately* from the structure.

You don't even need to understand how pattern matching works, for example, to implement this. Just follow the rules. (Though understanding is useful.)

To implement the right-hand side of a structural recursion (the `???`) the "following the types" strategy is almost always all that is needed.

Reply:
yup, I misunderstood this. I'll redo and we can review. Thanks for feedback!